File: ftl/samples/c/tibrequest.c
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample C request program which is used with the tibreply program
 * to demonstrate the basic use of inboxes.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "tib/ftl.h"

#include "tibaux.h"

#define DEFAULT_URL "http://localhost:8080"

const char*     realmServer             = DEFAULT_URL;
const char*     appName                 = "tibrequest";
const char*     endpointName            = "tibrequest-endpoint";
const char      *trcLevel               = NULL;
const char      *user                   = "guest";
const char      *password               = "guest-pw";
const char      *identifier             = NULL;
const char      *secondary              = NULL;
const char      *clientLabel            = NULL;

tibbool_t       replyReceived           = tibfalse;
tibdouble_t     pingInterval            = 1.0;
tibPublisher    pub                     = NULL;

volatile tibTimer   pongTimer           = NULL;

void
usage(int argc, char** argv)
{
    printf("Usage: %s [options] url\n", argv[0]);
    printf("Default url is "DEFAULT_URL"\n");
    printf("   -a, --application <name>\n"
           "       --client-label <label>\n"
           "   -e, --endpoint <name>\n"
           "   -h, --help\n"
           "   -id,--identifier\n"
           "   -p, --password <string>\n"
           "       --ping <float> (in seconds)\n"
           "   -s, --secondary <string>\n"
           "   -t, --trace <level>\n"
           "          where <level> can be:\n"
           "                off, severe, warn, info, debug, verbose\n"
           "   -u, --user <string>\n");

    exit(1);
}

void
parseArgs(int argc, char** argv)
{
    int     i;

    tibAux_InitGetArgs(argc, argv, usage);

    for (i = 1;  i < argc;  i++)
    {
        if (tibAux_GetString(&i, "--application", "-a", &appName))
            continue;
        if (tibAux_GetString(&i, "--client-label", "--client-label", &clientLabel))
            continue;
        if (tibAux_GetString(&i, "--endpoint", "-e", &endpointName))
            continue;
        if (tibAux_GetFlag(&i, "--help", "-h"))
            usage(argc, argv);
        if (tibAux_GetString(&i, "--identifier", "-id", &identifier))
            continue;
        if (tibAux_GetString(&i, "--password", "-p", &password))
            continue;
        if (tibAux_GetDouble(&i, "--ping", "--ping", &pingInterval))
            continue;
        if (tibAux_GetString(&i, "--secondary", "-s", &secondary))
            continue;
        if (tibAux_GetString(&i, "--trace", "-t", &trcLevel))
            continue;
        if (tibAux_GetString(&i, "--user", "-u", &user))
            continue;
        if (strncmp(argv[i], "-", 1)==0)
        {
            printf("invalid option: %s\n", argv[i]);
            usage(argc, argv);
        }
        realmServer = argv[i];
    }

    printf("Invoked as: ");
    for (i = 0;  i < argc;  i++)
        printf("%s ", argv[i]);
    printf("\n");
}

void
printMessage(
    tibEx               ex,
    tibMessage          msg)
{
    tibint32_t          len = 0;
    char                *buffer = 0;

    if (tibEx_GetErrorCode(ex) != TIB_OK)
        return;

    // get the size of output string
    len = tibMessage_ToString(ex, msg, NULL, 0);
    buffer = (char*)malloc(len);
    if (buffer)
    {
        (void)tibMessage_ToString(ex, msg, buffer, len);
        printf("  %s\n", buffer);
        free(buffer);
    }
}

void
onMsg(
    tibEx                       ex,
    tibEventQueue               msgQueue,
    tibint32_t                  msgNum,
    tibMessage                  *msgs,
    void                        **closures)
{
    if (msgNum != 1)
    {
        printf("Received %d messages instead of one.\n", msgNum);
        exit(1);
    }

    if (pongTimer)
    {
        tibEventQueue_DestroyTimer(ex, msgQueue, pongTimer, NULL);
        pongTimer = NULL;
    }

    replyReceived = tibtrue;

    printf("received reply message:\n");
    printMessage(ex, msgs[0]);

    // any exception that happens in the callback does not
    // propagate to the caller of the tibEventQueue_Dispatch
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
}

static void
onPongTimeout(
    tibEx            e,
    tibEventQueue    msgQueue,
    tibTimer         timer,
    void*            closure)
{
    tibMessage       ping = (tibMessage)closure;

    // Resend the initial ping message if we haven't received a pong.
    if (pongTimer)
    {
        if (trcLevel)
            printf("Resending initial message\n");
        tibPublisher_Send(e, pub, ping);
    }
}

int
main(int argc, char** argv)
{
    tibRealm                realm   = NULL;
    tibEventQueue           queue   = NULL;
    tibProperties           props   = NULL;
    tibSubscriber           sub     = NULL;
    tibInbox                inbox   = NULL;
    tibMessage              msg     = NULL;
    tibEx                   ex      = tibEx_Create();

    printf("#\n# %s\n#\n# %s\n#\n", argv[0], tib_Version());

    parseArgs(argc, argv);

    tib_Open(ex, TIB_COMPATIBILITY_VERSION);

    // Set global trace to specified level
    if (trcLevel != NULL)
        tib_SetLogLevel(ex, trcLevel);

    // Get a single realm per process
    props = tibProperties_Create(ex);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERNAME, user);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERPASSWORD, password);
    if (identifier)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
    if (secondary)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_SECONDARY_SERVER, secondary);

    // set the client label
    if (clientLabel != NULL)
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, clientLabel);
    else
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, argv[0]);

    realm = tibRealm_Connect(ex, realmServer, appName, props);
    tibProperties_Destroy(ex, props);

    queue = tibEventQueue_Create(ex, realm, NULL);

    sub = tibSubscriber_CreateOnInbox(ex, realm, endpointName, NULL);
    inbox = tibSubscriber_GetInbox(ex, sub);
    tibEventQueue_AddSubscriber(ex, queue, sub, onMsg, NULL);

    pub = tibPublisher_Create(ex, realm, endpointName, NULL);

    msg = tibMessage_Create(ex, realm, "Format-1");

    // set by name since performance is not demonstrated here
    tibMessage_SetString(ex, msg, "My-String", "request");
    tibMessage_SetLong(ex, msg, "My-Long", 10);

    tibMessage_SetOpaque(ex, msg, "My-Opaque", "payload", sizeof("payload"));

    // put our inbox in the request messge
    // set by name since performance is not demonstrated here
    tibMessage_SetInbox(ex, msg, "My-Inbox", inbox);

    if (pingInterval)
        pongTimer = tibEventQueue_CreateTimer(ex, queue, pingInterval,
                                              onPongTimeout, msg);

    printf("sending request message:\n");
    printMessage(ex, msg);
    
    // send the request message
    tibPublisher_Send(ex, pub, msg);

    // wait for the reply message. it will be sent to our inbox
    while (tibEx_GetErrorCode(ex) == TIB_OK && !replyReceived)
        tibEventQueue_Dispatch(ex, queue, TIB_TIMEOUT_WAIT_FOREVER);

    // Clean up
    if (queue && sub)
        tibEventQueue_RemoveSubscriber(ex, queue, sub, NULL);
    if (pongTimer)
        tibEventQueue_DestroyTimer(ex, queue, pongTimer, NULL);
    tibEventQueue_Destroy(ex, queue, NULL);
    tibSubscriber_Close(ex, sub);
    tibPublisher_Close(ex, pub);
    tibMessage_Destroy(ex, msg);
    tibRealm_Close(ex, realm);
    tib_Close(ex);

    // Check if it worked before destroying the exception
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
    tibEx_Destroy(ex);

    return 0;
}

File: ftl/samples/java/TibRequest.java
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample Java request program which is used with the TibReply program
 * to demonstrate the basic use of inboxes.
 */

package com.tibco.ftl.samples;

import com.tibco.ftl.*;

import java.util.*;

public class TibRequest implements SubscriberListener, EventTimerListener
{
    String          appName       = "tibrequest";
    String          realmServer   = "http://localhost:8080";
    String          endpointName  = "tibrequest-endpoint";
    String          trcLevel      = null;
    String          user          = "guest";
    String          password      = "guest-pw";
    String          identifier    = null;
    String          secondary     = null;
    String          clientLabel   = null;

    double          pingInterval  = 1.0;
    volatile EventTimer pongTimer = null;
    Message         msg;
    Publisher       pub;

    boolean         replyReceived = false;

    String          usageString =
        "TibRequest [options] url\n" +
        "Default url is " + realmServer + "\n" +
        "  where options can be:\n" +
        "    -a, --application <name>\n" +
        "    -cl, --client-label <string>\n" +
        "    -e, --endpoint <name>\n" +
        "    -h, --help\n" +
        "    -id,--identifier <string>\n" +
        "    -p, --password <string>\n" +
        "        --ping <float> (in seconds)\n" +
        "    -s, --secondary <string>\n" +
        "    -t, --trace <level>\n" +
        "           where <level> can be:\n" +
        "                 off, severe, warn, info, debug, verbose\n" +
        "    -u, --user <string>\n";

    public TibRequest(String[] args)
    {
        System.out.printf("#\n# %s\n#\n# %s\n#\n",
                          this.getClass().getName(),
                          FTL.getVersionInformation());
        
        parseArgs(args);
    }

    public void usage()
    {
        System.out.println(usageString);
        System.exit(0);
    }

    public void parseArgs(String args[])
    {
        int i           = 0;
        double d        = 0;
        int argc        = args.length;
        String s        = null;
        TibAux aux      = new TibAux(args, usageString);

        for (i = 0; i < argc; i++)
        {
            s = aux.getString(i, "--application", "-a");
            if (s != null) {
                appName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--client-label", "-cl");
            if (s != null) {
                clientLabel = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--endpoint", "-e");
            if (s != null) {
                endpointName = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--help", "-h")) {
                usage();
            }
            s = aux.getString(i, "--identifier", "-id");
            if (s != null) {
                identifier = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--password", "-p");
            if (s != null) {
                password = s;
                i++;
                continue;
            }
            d = aux.getDouble(i, "--ping", "--ping");
            if (d > -1.0) {
                pingInterval = d;
                i++;
                continue;
            }
            s = aux.getString(i, "--secondary", "-s");
            if (s != null) {
                secondary = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--trace", "-t");
            if (s != null) {
                trcLevel = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--user", "-u");
            if (s != null) {
                user = s;
                i++;
                continue;
            }
            if (args[i].startsWith("-", 0)) {
                System.out.println("invalid option: "+args[i]);
                usage();
            }
            realmServer = args[i];
        }

        System.out.print("Invoked as: TibRequest ");
        for (i = 0; i < argc; i++)
            System.out.print(args[i]+" ");
        System.out.println("");
    }

    public void messagesReceived(List<Message> messages, EventQueue eventQueue)
    {
        try
        {
            if (pongTimer != null)
            {
                eventQueue.destroyTimer(pongTimer);
                pongTimer = null;
            }
        }
        catch (FTLException e)
        {
            e.printStackTrace();
        }
        replyReceived = true;
        System.out.printf("received reply message:\n");
        System.out.println("  " + messages.get(0));
    }

    public void timerFired(EventTimer timer,EventQueue eventQueue)
    {
        try
        {
            if (trcLevel != null)
                System.out.printf("resending initial message\n");
            pub.send(msg);
        }
        catch (FTLException e)
        {
            e.printStackTrace();
            replyReceived = true;
        }
    }

    public int request() throws FTLException
    {
        TibProperties   props;
        Realm           realm;
        EventQueue      queue;
        InboxSubscriber sub;
        Inbox           inbox;

        // Set global trace to specified level
        if (trcLevel != null)
            FTL.setLogLevel(trcLevel);

        // Get a single realm per process
        props = FTL.createProperties();
        props.set(Realm.PROPERTY_STRING_USERNAME, user);
        props.set(Realm.PROPERTY_STRING_USERPASSWORD, password);
        if (identifier != null)
            props.set(Realm.PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
        if (secondary != null)
            props.set(Realm.PROPERTY_STRING_SECONDARY_SERVER, secondary);

        if (clientLabel != null)
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, clientLabel);
        else
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, this.getClass().getSimpleName());

        realm = FTL.connectToRealmServer(realmServer, appName, props);
        props.destroy();

        sub = realm.createInboxSubscriber(endpointName);
        inbox = sub.getInbox();
        queue = realm.createEventQueue();
        queue.addSubscriber(sub, this);

        pub = realm.createPublisher(endpointName);
        msg = realm.createMessage("Format-1");

        // set by name since performance is not demonstrated here
        msg.setString("My-String", "request");
        msg.setLong("My-Long", 10);

        byte[] opaqueBytes = (new String("payload")).getBytes(); 
        msg.setOpaque("My-Opaque", opaqueBytes);

        // put our inbox in the request message
        // set by name since performance is not demonstrated here
        msg.setInbox("My-Inbox", inbox);

        // Should we retry the initial ping until we get a pong?
        if (pingInterval != 0)
            pongTimer = queue.createTimer(pingInterval, this);

        System.out.printf("sending request message:\n");
        System.out.println("  "  + msg);

        // send the request message
        pub.send(msg);

        while (!replyReceived)
            queue.dispatch();

        // Clean up
        queue.removeSubscriber(sub);
        if (pongTimer != null)
            queue.destroyTimer(pongTimer);
        queue.destroy();
        sub.close();
        pub.close();
        msg.destroy();
        realm.close();

        return 0;
    }
    
    public static void main(String[] args)
    {
        TibRequest s  = new TibRequest(args);
        try
        {
            s.request();
        }
        catch (Throwable e)
        {
            e.printStackTrace();
        }
    }
}
